File: nodes/get-messages-range/get-messages-range.process.py
from typing import Any, Dict, List, Union

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    try:
        messages = inputs.get('messages')
        start = inputs.get('start')
        end = inputs.get('end')

        if not isinstance(messages, list):
            raise Exception("Messages input must be an array")

        if not isinstance(start, int) or start < 0:
            raise Exception("Start index must be a non-negative integer")

        if not isinstance(end, int):
            raise Exception("End index must be an integer")

        # Handle -1 as end of array
        if end == -1:
            end = len(messages)

        if end < start:
            raise Exception("End index must be greater than or equal to start index")

        # Clamp indices to array bounds
        start = max(0, min(start, len(messages)))
        end = max(start, min(end, len(messages)))

        # Extract messages in range
        range_messages = messages[start:end]

        return {
            "messages": range_messages,
            "count": len(range_messages)
        }

    except Exception as error:
        print(f'Get Messages Range error: {error}')
        raise Exception(f"Get Messages Range error: {str(error)}")

File: sdks/nodejs/nodes/get-messages-range/get-messages-range.process.js
export default async ({inputs, settings, config}) => {
    try {
        const messages = inputs.messages;
        let start = inputs.start;
        let end = inputs.end;

        if (!Array.isArray(messages)) {
            throw new Error("Messages input must be an array");
        }

        if (typeof start !== 'number' || !Number.isInteger(start) || start < 0) {
            throw new Error("Start index must be a non-negative integer");
        }

        if (typeof end !== 'number' || !Number.isInteger(end)) {
            throw new Error("End index must be an integer");
        }

        // Handle -1 as end of array
        if (end === -1) {
            end = messages.length;
        }

        if (end < start) {
            throw new Error("End index must be greater than or equal to start index");
        }

        // Clamp indices to array bounds
        start = Math.max(0, Math.min(start, messages.length));
        end = Math.max(start, Math.min(end, messages.length));

        // Extract messages in range
        const rangeMessages = messages.slice(start, end);

        return {
            messages: rangeMessages,
            count: rangeMessages.length
        };

    } catch (error) {
        console.log('Get Messages Range error:', error);
        throw new Error(`Get Messages Range error: ${error.message}`);
    }
};
